### PR TITLE
Support withdrawals in t8n

### PR DESCRIPTION
**Pull request by rodiazet**
- Add support for withdrawals in t8n interface. It's required by `execution-spec-tests`
 - Tests

### CODE REVIEW

File: test/t8n/t8n.cpp [hunk 1]
```diff
@@ -160,6 +160,10 @@ int main(int argc, const char* argv[])
             if (block_reward.has_value())
                 state.touch(block.coinbase).balance += *block_reward;
 
+            // Apply withdrawals. Amount value is in gwei.
+            for (const auto& withdraw : block.withdrawals)
+                state.touch(withdraw.first).balance += withdraw.second * 1000000000;
```
Comment (chfast) [inline test/t8n/t8n.cpp:165]: Are withdrawals applies before block finalization? I.e. are touched accounts possibly removed?

File: test/state/state.hpp [hunk 2]
```diff
@@ -66,6 +66,8 @@ class State
     [[nodiscard]] const auto& get_accounts() const noexcept { return m_accounts; }
 };
 
+using Withdrawals = std::vector<std::pair<address, intx::uint256>>;
```
Comment (chfast) [inline test/state/state.hpp:69]: The spec says the amount is `uint64`.

**Review by chfast — CHANGES_REQUESTED** [review]

File: test/t8n/t8n.cpp (hunk 1 above)
Comment (rodiazet) [reply test/t8n/t8n.cpp:165]: Spec says that `amount` is non-zero. So I will not be removed on finalizing. Am I right? https://eips.ethereum.org/EIPS/eip-4895

File: test/state/state.hpp (hunk 2 above)
Comment (rodiazet) [reply test/state/state.hpp]: Done

File: test/state/state.hpp [hunk 3]
```diff
@@ -66,6 +66,8 @@ class State
     [[nodiscard]] const auto& get_accounts() const noexcept { return m_accounts; }
 };
 
+using Withdrawals = std::vector<std::pair<address, uint64_t>>;
```
Comment (axic) [inline test/state/state.hpp:69]: Not necessary but perhaps a comment explaining what the value is and what range it has is useful.

Comment (rodiazet) [reply test/state/state.hpp:69]: done

Comment (chfast) [reply test/state/state.hpp:69]: I think better will be to just define single `using Withdrawal = std::pair<address, uint64_t>`. But at this point a struct with named fields will do better.

File: test/statetest/statetest_loader.cpp [hunk 4]
```diff
@@ -170,9 +170,20 @@ state::BlockInfo from_json<state::BlockInfo>(const json::json& j)
...
+        {
+            withdrawals.push_back({from_json<evmc::address>(withdraw.at("address")),
+                from_json<uint64_t>(withdraw.at("amount"))});
+        }
+    }
+
     return {from_json<int64_t>(j.at("currentNumber")), from_json<int64_t>(j.at("currentTimestamp")),
         from_json<int64_t>(j.at("currentGasLimit")),
-        from_json<evmc::address>(j.at("currentCoinbase")), difficulty, base_fee};
+        from_json<evmc::address>(j.at("currentCoinbase")), difficulty, base_fee, withdrawals};
```
Comment (chfast) [inline test/statetest/statetest_loader.cpp:186]: move

File: test/t8n/t8n.cpp [hunk 5]
```diff
@@ -159,6 +159,10 @@ int main(int argc, const char* argv[])
 
             state::finalize(state, rev, block.coinbase, block_reward);
 
+            // Apply withdrawals. Amount value is in gwei.
+            for (const auto& withdraw : block.withdrawals)
```
Comment (chfast) [inline test/t8n/t8n.cpp:163]: You should pass it to `finalize()` as `std::span<Withdrawal>`.

File: test/state/state.cpp [hunk 6]
```diff
@@ -131,6 +131,10 @@ void finalize(
                 return acc.erasable && acc.is_empty();
             });
     }
+
+    // Apply withdrawals. Amount value is in gwei.
+    for (const auto& withdraw : withdrawals)
+        state.touch(withdraw.first).balance += intx::uint256{withdraw.second} * 1000000000;
```
Comment (axic) [inline test/state/state.cpp:137]: Not sure if this is better but avoids counting zeroes:
```suggestion
        state.touch(withdraw.first).balance += intx::uint256{withdraw.second} * 1e9;
```

Not sure what is our preference in the codebase, @chfast.

File: test/state/state.hpp [hunk 7]
```diff
@@ -121,9 +126,10 @@ struct TransactionReceipt
 
 /// Finalize state after applying a "block" of transactions.
 ///
-/// Applies block reward to coinbase and deletes empty touched accounts (post Spurious Dragon).
-void finalize(
-    State& state, evmc_revision rev, const address& coinbase, std::optional<uint64_t> block_reward);
+/// Applies block reward to coinbase, withdrawals and deletes empty touched accounts (post Spurious
```
Comment (axic) [inline test/state/state.hpp:129]:
```suggestion
/// Applies block reward to coinbase, withdrawals (post Shanghai) and deletes empty touched accounts (post Spurious
```

File: test/unittests/statetest_loader_block_info_test.cpp [hunk 8]
```diff
@@ -170,3 +170,43 @@ TEST(statetest_loader, block_info_0_random)
...
+    EXPECT_EQ(bi.prev_randao, 0x00_bytes32);
+    EXPECT_EQ(bi.gas_limit, 0x0);
+    EXPECT_EQ(bi.base_fee, 7);
+    EXPECT_EQ(bi.timestamp, 0);
+    EXPECT_EQ(bi.number, 0);
+    EXPECT_EQ(bi.withdrawals.size(), 2);
+    EXPECT_EQ(bi.withdrawals[0].recipient, 0x0000000000000000000000000000000000000100_address);
+    EXPECT_EQ(bi.withdrawals[0].amount, 0x800000000);
+    EXPECT_EQ(bi.withdrawals[1].recipient, 0x0000000000000000000000000000000000000200_address);
+    EXPECT_EQ(bi.withdrawals[1].amount, 0xffffffffffffffff);
```
Comment (axic) [inline test/unittests/statetest_loader_block_info_test.cpp:211]: If we have the `amount_wei` helper, could also check for the output that here.

Comment (rodiazet) [reply test/unittests/statetest_loader_block_info_test.cpp]: done

File: test/state/state.hpp [hunk 9]
```diff
@@ -66,6 +66,20 @@ class State
     [[nodiscard]] const auto& get_accounts() const noexcept { return m_accounts; }
 };
 
+struct Withdrawal
+{
+    address recipient;
+    /// The amount is denominated in gwei.
```
Comment (chfast) [inline test/state/state.hpp:72]:
```suggestion

    /// The amount is denominated in gwei.
```
or use `///<`.

**Review by chfast — CHANGES_REQUESTED** [review]

**Review by chfast — APPROVED** [review]
Please squash.

**Review by axic — APPROVED** [review]
Looks good to me but please address @chfast's comments first.

**Pull request merged**